mod-source-record-storage-server/src/main/resources/liquibase/tenant/scripts/v-5.10.0/2024-12-10--20-00-create-delete-old-marc-indexers-versions.xml

```diff
@@ -0,0 +1,78 @@
+<databaseChangeLog
+  xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+    http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.0.xsd">
+
+
+  <changeSet id="2024-12-10--17-00-update-function-insert_marc_indexers" author="Volodymyr_Rohach">
+    <sql splitStatements="false">
+      create or replace function ${database.defaultSchemaName}.insert_marc_indexers()
+      returns trigger
+      as
+      $insert_marc_indexers$
+      declare
+      result integer;
+      begin
+          if (TG_OP = 'INSERT') then
+              insert into ${database.defaultSchemaName}.marc_records_tracking(marc_id, version, is_dirty) values (NEW.id, 0, false);
+      result = 0;
+          elsif (TG_OP = 'UPDATE') then
+      update ${database.defaultSchemaName}.marc_records_tracking set version = version + 1,
+                                                                     is_dirty = true
+      where marc_id = NEW.id
+        returning version into result;
+      delete from ${database.defaultSchemaName}.marc_indexers_leader where marc_id = NEW.id;
+      end if;
+          perform ${database.defaultSchemaName}.fill_in_marc_indexers(NEW.id, NEW.content, result);
+      return NEW;
+      end;
+      $insert_marc_indexers$ language plpgsql;
+    </sql>
+  </changeSet>
+
+  <changeSet id="2024-12-10--17-00-create-delete-old-marc-indexers-versions" author="Volodymyr_Rohach">
+    <createProcedure>
+      CREATE OR REPLACE PROCEDURE ${database.defaultSchemaName}.delete_old_marc_indexers_versions() LANGUAGE plpgsql AS $$
+      BEGIN
+      WITH dirty_records AS (
+        SELECT mrt.marc_id, mrt.version
+        FROM marc_records_tracking mrt
+        WHERE mrt.is_dirty = true
+      ),
+      deleted_dirty AS (
+      DELETE FROM marc_indexers mi
+        USING dirty_records dr
+      WHERE mi.marc_id = dr.marc_id AND dr.version > mi.version
+        RETURNING mi.marc_id
+        ),
+      old_records AS (
+      SELECT ot.marc_id
+      FROM old_records_tracking ot
+      WHERE ot.has_been_processed = false
+        ),
+      deleted_old AS (
+      DELETE FROM marc_indexers mi
+        USING old_records ors
+      WHERE mi.marc_id = ors.marc_id
+        RETURNING mi.marc_id
+        ),
+      combined_deletions AS (
+      SELECT marc_id FROM deleted_dirty
+      UNION
+      SELECT marc_id FROM deleted_old
+        )
+      INSERT INTO marc_indexers_deleted_ids (marc_id)
+      SELECT DISTINCT marc_id FROM combined_deletions
+        ON CONFLICT (marc_id) DO NOTHING;
+      END $$;
+    </createProcedure>
+  </changeSet>
+
+
+  <changeSet id="2024-12-10--17-00-create-index-on-is-dirty-true" author="Volodymyr_Rohach">
+    <sql>
+      CREATE INDEX IF NOT EXISTS idx_marc_records_tracking_dirty_true ON ${database.defaultSchemaName}.marc_records_tracking (is_dirty) WHERE is_dirty = true;
```

Comment:
i think this index should be is_dirty, marc_id, version and maybe still add the where clause. This is support the `dirty_records` CTE defined in the delete procedure. The current index still helps, and it could go further by not reaching out to the table to get marc_id and version.

Reply:
@okolawole-ebsco Thank you! Done.